// @shoelace-style/shoelace/dist/chunks/chunk.2VBMOEKG.js
// node_modules/lit-html/lit-html.js
var t;
var i = globalThis.trustedTypes;
var s = i ? i.createPolicy("lit-html", { createHTML: (t4) => t4 }) : void 0;
var e = `lit$${(Math.random() + "").slice(9)}$`;
var o = "?" + e;
var n = `<${o}>`;
var l = document;
var h = (t4 = "") => l.createComment(t4);
var r = (t4) => t4 === null || typeof t4 != "object" && typeof t4 != "function";
var d = Array.isArray;
var u = (t4) => {
  var i4;
  return d(t4) || typeof ((i4 = t4) === null || i4 === void 0 ? void 0 : i4[Symbol.iterator]) == "function";
};
var c = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g;
var v = /-->/g;
var a = />/g;
var f = />|[ 	\n\r](?:([^\s"'>=/]+)([ 	\n\r]*=[ 	\n\r]*(?:[^ 	\n\r"'`<>=]|("|')|))|$)/g;
var _ = /'/g;
var m = /"/g;
var g = /^(?:script|style|textarea)$/i;
var p = (t4) => (i4, ...s7) => ({ _$litType$: t4, strings: i4, values: s7 });
var $ = p(1);
var y = p(2);
var b = Symbol.for("lit-noChange");
var w = Symbol.for("lit-nothing");
var T = /* @__PURE__ */ new WeakMap();
var x = (t4, i4, s7) => {
  var e6, o7;
  const n7 = (e6 = s7 == null ? void 0 : s7.renderBefore) !== null && e6 !== void 0 ? e6 : i4;
  let l5 = n7._$litPart$;
  if (l5 === void 0) {
    const t5 = (o7 = s7 == null ? void 0 : s7.renderBefore) !== null && o7 !== void 0 ? o7 : null;
    n7._$litPart$ = l5 = new N(i4.insertBefore(h(), t5), t5, void 0, s7 != null ? s7 : {});
  }
  return l5._$AI(t4), l5;
};
var A = l.createTreeWalker(l, 129, null, false);
var C = (t4, i4) => {
  const o7 = t4.length - 1, l5 = [];
  let h4, r6 = i4 === 2 ? "<svg>" : "", d2 = c;
  for (let i5 = 0; i5 < o7; i5++) {
    const s7 = t4[i5];
    let o8, u3, p2 = -1, $2 = 0;
    for (; $2 < s7.length && (d2.lastIndex = $2, u3 = d2.exec(s7), u3 !== null); )
      $2 = d2.lastIndex, d2 === c ? u3[1] === "!--" ? d2 = v : u3[1] !== void 0 ? d2 = a : u3[2] !== void 0 ? (g.test(u3[2]) && (h4 = RegExp("</" + u3[2], "g")), d2 = f) : u3[3] !== void 0 && (d2 = f) : d2 === f ? u3[0] === ">" ? (d2 = h4 != null ? h4 : c, p2 = -1) : u3[1] === void 0 ? p2 = -2 : (p2 = d2.lastIndex - u3[2].length, o8 = u3[1], d2 = u3[3] === void 0 ? f : u3[3] === '"' ? m : _) : d2 === m || d2 === _ ? d2 = f : d2 === v || d2 === a ? d2 = c : (d2 = f, h4 = void 0);
    const y2 = d2 === f && t4[i5 + 1].startsWith("/>") ? " " : "";
    r6 += d2 === c ? s7 + n : p2 >= 0 ? (l5.push(o8), s7.slice(0, p2) + "$lit$" + s7.slice(p2) + e + y2) : s7 + e + (p2 === -2 ? (l5.push(void 0), i5) : y2);
  }
  const u2 = r6 + (t4[o7] || "<?>") + (i4 === 2 ? "</svg>" : "");
  if (!Array.isArray(t4) || !t4.hasOwnProperty("raw"))
    throw Error("invalid template strings array");
  return [s !== void 0 ? s.createHTML(u2) : u2, l5];
};
var E = class {
  constructor({ strings: t4, _$litType$: s7 }, n7) {
    let l5;
    this.parts = [];
    let r6 = 0, d2 = 0;
    const u2 = t4.length - 1, c2 = this.parts, [v2, a4] = C(t4, s7);
    if (this.el = E.createElement(v2, n7), A.currentNode = this.el.content, s7 === 2) {
      const t5 = this.el.content, i4 = t5.firstChild;
      i4.remove(), t5.append(...i4.childNodes);
    }
    for (; (l5 = A.nextNode()) !== null && c2.length < u2; ) {
      if (l5.nodeType === 1) {
        if (l5.hasAttributes()) {
          const t5 = [];
          for (const i4 of l5.getAttributeNames())
            if (i4.endsWith("$lit$") || i4.startsWith(e)) {
              const s8 = a4[d2++];
              if (t5.push(i4), s8 !== void 0) {
                const t6 = l5.getAttribute(s8.toLowerCase() + "$lit$").split(e), i5 = /([.?@])?(.*)/.exec(s8);
                c2.push({ type: 1, index: r6, name: i5[2], strings: t6, ctor: i5[1] === "." ? M : i5[1] === "?" ? H : i5[1] === "@" ? I : S });
              } else
                c2.push({ type: 6, index: r6 });
            }
          for (const i4 of t5)
            l5.removeAttribute(i4);
        }
        if (g.test(l5.tagName)) {
          const t5 = l5.textContent.split(e), s8 = t5.length - 1;
          if (s8 > 0) {
            l5.textContent = i ? i.emptyScript : "";
            for (let i4 = 0; i4 < s8; i4++)
              l5.append(t5[i4], h()), A.nextNode(), c2.push({ type: 2, index: ++r6 });
            l5.append(t5[s8], h());
          }
        }
      } else if (l5.nodeType === 8)
        if (l5.data === o)
          c2.push({ type: 2, index: r6 });
        else {
          let t5 = -1;
          for (; (t5 = l5.data.indexOf(e, t5 + 1)) !== -1; )
            c2.push({ type: 7, index: r6 }), t5 += e.length - 1;
        }
      r6++;
    }
  }
  static createElement(t4, i4) {
    const s7 = l.createElement("template");
    return s7.innerHTML = t4, s7;
  }
};
function P(t4, i4, s7 = t4, e6) {
  var o7, n7, l5, h4;
  if (i4 === b)
    return i4;
  let d2 = e6 !== void 0 ? (o7 = s7._$Cl) === null || o7 === void 0 ? void 0 : o7[e6] : s7._$Cu;
  const u2 = r(i4) ? void 0 : i4._$litDirective$;
  return (d2 == null ? void 0 : d2.constructor) !== u2 && ((n7 = d2 == null ? void 0 : d2._$AO) === null || n7 === void 0 || n7.call(d2, false), u2 === void 0 ? d2 = void 0 : (d2 = new u2(t4), d2._$AT(t4, s7, e6)), e6 !== void 0 ? ((l5 = (h4 = s7)._$Cl) !== null && l5 !== void 0 ? l5 : h4._$Cl = [])[e6] = d2 : s7._$Cu = d2), d2 !== void 0 && (i4 = P(t4, d2._$AS(t4, i4.values), d2, e6)), i4;
}
var V = class {
  constructor(t4, i4) {
    this.v = [], this._$AN = void 0, this._$AD = t4, this._$AM = i4;
  }
  get parentNode() {
    return this._$AM.parentNode;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  p(t4) {
    var i4;
    const { el: { content: s7 }, parts: e6 } = this._$AD, o7 = ((i4 = t4 == null ? void 0 : t4.creationScope) !== null && i4 !== void 0 ? i4 : l).importNode(s7, true);
    A.currentNode = o7;
    let n7 = A.nextNode(), h4 = 0, r6 = 0, d2 = e6[0];
    for (; d2 !== void 0; ) {
      if (h4 === d2.index) {
        let i5;
        d2.type === 2 ? i5 = new N(n7, n7.nextSibling, this, t4) : d2.type === 1 ? i5 = new d2.ctor(n7, d2.name, d2.strings, this, t4) : d2.type === 6 && (i5 = new L(n7, this, t4)), this.v.push(i5), d2 = e6[++r6];
      }
      h4 !== (d2 == null ? void 0 : d2.index) && (n7 = A.nextNode(), h4++);
    }
    return o7;
  }
  m(t4) {
    let i4 = 0;
    for (const s7 of this.v)
      s7 !== void 0 && (s7.strings !== void 0 ? (s7._$AI(t4, s7, i4), i4 += s7.strings.length - 2) : s7._$AI(t4[i4])), i4++;
  }
};
var N = class {
  constructor(t4, i4, s7, e6) {
    var o7;
    this.type = 2, this._$AH = w, this._$AN = void 0, this._$AA = t4, this._$AB = i4, this._$AM = s7, this.options = e6, this._$Cg = (o7 = e6 == null ? void 0 : e6.isConnected) === null || o7 === void 0 || o7;
  }
  get _$AU() {
    var t4, i4;
    return (i4 = (t4 = this._$AM) === null || t4 === void 0 ? void 0 : t4._$AU) !== null && i4 !== void 0 ? i4 : this._$Cg;
  }
  get parentNode() {
    let t4 = this._$AA.parentNode;
    const i4 = this._$AM;
    return i4 !== void 0 && t4.nodeType === 11 && (t4 = i4.parentNode), t4;
  }
  get startNode() {
    return this._$AA;
  }
  get endNode() {
    return this._$AB;
  }
  _$AI(t4, i4 = this) {
    t4 = P(this, t4, i4), r(t4) ? t4 === w || t4 == null || t4 === "" ? (this._$AH !== w && this._$AR(), this._$AH = w) : t4 !== this._$AH && t4 !== b && this.$(t4) : t4._$litType$ !== void 0 ? this.T(t4) : t4.nodeType !== void 0 ? this.S(t4) : u(t4) ? this.A(t4) : this.$(t4);
  }
  M(t4, i4 = this._$AB) {
    return this._$AA.parentNode.insertBefore(t4, i4);
  }
  S(t4) {
    this._$AH !== t4 && (this._$AR(), this._$AH = this.M(t4));
  }
  $(t4) {
    this._$AH !== w && r(this._$AH) ? this._$AA.nextSibling.data = t4 : this.S(l.createTextNode(t4)), this._$AH = t4;
  }
  T(t4) {
    var i4;
    const { values: s7, _$litType$: e6 } = t4, o7 = typeof e6 == "number" ? this._$AC(t4) : (e6.el === void 0 && (e6.el = E.createElement(e6.h, this.options)), e6);
    if (((i4 = this._$AH) === null || i4 === void 0 ? void 0 : i4._$AD) === o7)
      this._$AH.m(s7);
    else {
      const t5 = new V(o7, this), i5 = t5.p(this.options);
      t5.m(s7), this.S(i5), this._$AH = t5;
    }
  }
  _$AC(t4) {
    let i4 = T.get(t4.strings);
    return i4 === void 0 && T.set(t4.strings, i4 = new E(t4)), i4;
  }
  A(t4) {
    d(this._$AH) || (this._$AH = [], this._$AR());
    const i4 = this._$AH;
    let s7, e6 = 0;
    for (const o7 of t4)
      e6 === i4.length ? i4.push(s7 = new N(this.M(h()), this.M(h()), this, this.options)) : s7 = i4[e6], s7._$AI(o7), e6++;
    e6 < i4.length && (this._$AR(s7 && s7._$AB.nextSibling, e6), i4.length = e6);
  }
  _$AR(t4 = this._$AA.nextSibling, i4) {
    var s7;
    for ((s7 = this._$AP) === null || s7 === void 0 || s7.call(this, false, true, i4); t4 && t4 !== this._$AB; ) {
      const i5 = t4.nextSibling;
      t4.remove(), t4 = i5;
    }
  }
  setConnected(t4) {
    var i4;
    this._$AM === void 0 && (this._$Cg = t4, (i4 = this._$AP) === null || i4 === void 0 || i4.call(this, t4));
  }
};
var S = class {
  constructor(t4, i4, s7, e6, o7) {
    this.type = 1, this._$AH = w, this._$AN = void 0, this.element = t4, this.name = i4, this._$AM = e6, this.options = o7, s7.length > 2 || s7[0] !== "" || s7[1] !== "" ? (this._$AH = Array(s7.length - 1).fill(new String()), this.strings = s7) : this._$AH = w;
  }
  get tagName() {
    return this.element.tagName;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t4, i4 = this, s7, e6) {
    const o7 = this.strings;
    let n7 = false;
    if (o7 === void 0)
      t4 = P(this, t4, i4, 0), n7 = !r(t4) || t4 !== this._$AH && t4 !== b, n7 && (this._$AH = t4);
    else {
      const e7 = t4;
      let l5, h4;
      for (t4 = o7[0], l5 = 0; l5 < o7.length - 1; l5++)
        h4 = P(this, e7[s7 + l5], i4, l5), h4 === b && (h4 = this._$AH[l5]), n7 || (n7 = !r(h4) || h4 !== this._$AH[l5]), h4 === w ? t4 = w : t4 !== w && (t4 += (h4 != null ? h4 : "") + o7[l5 + 1]), this._$AH[l5] = h4;
    }
    n7 && !e6 && this.k(t4);
  }
  k(t4) {
    t4 === w ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, t4 != null ? t4 : "");
  }
};
var M = class extends S {
  constructor() {
    super(...arguments), this.type = 3;
  }
  k(t4) {
    this.element[this.name] = t4 === w ? void 0 : t4;
  }
};
var k = i ? i.emptyScript : "";
var H = class extends S {
  constructor() {
    super(...arguments), this.type = 4;
  }
  k(t4) {
    t4 && t4 !== w ? this.element.setAttribute(this.name, k) : this.element.removeAttribute(this.name);
  }
};
var I = class extends S {
  constructor(t4, i4, s7, e6, o7) {
    super(t4, i4, s7, e6, o7), this.type = 5;
  }
  _$AI(t4, i4 = this) {
    var s7;
    if ((t4 = (s7 = P(this, t4, i4, 0)) !== null && s7 !== void 0 ? s7 : w) === b)
      return;
    const e6 = this._$AH, o7 = t4 === w && e6 !== w || t4.capture !== e6.capture || t4.once !== e6.once || t4.passive !== e6.passive, n7 = t4 !== w && (e6 === w || o7);
    o7 && this.element.removeEventListener(this.name, this, e6), n7 && this.element.addEventListener(this.name, this, t4), this._$AH = t4;
  }
  handleEvent(t4) {
    var i4, s7;
    typeof this._$AH == "function" ? this._$AH.call((s7 = (i4 = this.options) === null || i4 === void 0 ? void 0 : i4.host) !== null && s7 !== void 0 ? s7 : this.element, t4) : this._$AH.handleEvent(t4);
  }
};
var L = class {
  constructor(t4, i4, s7) {
    this.element = t4, this.type = 6, this._$AN = void 0, this._$AM = i4, this.options = s7;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t4) {
    P(this, t4);
  }
};
var z = window.litHtmlPolyfillSupport;
z == null || z(E, N), ((t = globalThis.litHtmlVersions) !== null && t !== void 0 ? t : globalThis.litHtmlVersions = []).push("2.1.0");

// node_modules/lit-element/node_modules/@lit/reactive-element/css-tag.js
var t2 = window.ShadowRoot && (window.ShadyCSS === void 0 || window.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype;
var e2 = Symbol();
var n2 = /* @__PURE__ */ new Map();
var s2 = class {
  constructor(t4, n7) {
    if (this._$cssResult$ = true, n7 !== e2)
      throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t4;
  }
  get styleSheet() {
    let e6 = n2.get(this.cssText);
    return t2 && e6 === void 0 && (n2.set(this.cssText, e6 = new CSSStyleSheet()), e6.replaceSync(this.cssText)), e6;
  }
  toString() {
    return this.cssText;
  }
};
var o2 = (t4) => new s2(typeof t4 == "string" ? t4 : t4 + "", e2);
var r2 = (t4, ...n7) => {
  const o7 = t4.length === 1 ? t4[0] : n7.reduce((e6, n8, s7) => e6 + ((t5) => {
    if (t5._$cssResult$ === true)
      return t5.cssText;
    if (typeof t5 == "number")
      return t5;
    throw Error("Value passed to 'css' function must be a 'css' function result: " + t5 + ". Use 'unsafeCSS' to pass non-literal values, but take care to ensure page security.");
  })(n8) + t4[s7 + 1], t4[0]);
  return new s2(o7, e2);
};
var i2 = (e6, n7) => {
  t2 ? e6.adoptedStyleSheets = n7.map((t4) => t4 instanceof CSSStyleSheet ? t4 : t4.styleSheet) : n7.forEach((t4) => {
    const n8 = document.createElement("style"), s7 = window.litNonce;
    s7 !== void 0 && n8.setAttribute("nonce", s7), n8.textContent = t4.cssText, e6.appendChild(n8);
  });
};
var S2 = t2 ? (t4) => t4 : (t4) => t4 instanceof CSSStyleSheet ? ((t5) => {
  let e6 = "";
  for (const n7 of t5.cssRules)
    e6 += n7.cssText;
  return o2(e6);
})(t4) : t4;

// node_modules/lit-element/node_modules/@lit/reactive-element/reactive-element.js
var s3;
var e3 = window.trustedTypes;
var r3 = e3 ? e3.emptyScript : "";
var h2 = window.reactiveElementPolyfillSupport;
var o3 = { toAttribute(t4, i4) {
  switch (i4) {
    case Boolean:
      t4 = t4 ? r3 : null;
      break;
    case Object:
    case Array:
      t4 = t4 == null ? t4 : JSON.stringify(t4);
  }
  return t4;
}, fromAttribute(t4, i4) {
  let s7 = t4;
  switch (i4) {
    case Boolean:
      s7 = t4 !== null;
      break;
    case Number:
      s7 = t4 === null ? null : Number(t4);
      break;
    case Object:
    case Array:
      try {
        s7 = JSON.parse(t4);
      } catch (t5) {
        s7 = null;
      }
  }
  return s7;
} };
var n3 = (t4, i4) => i4 !== t4 && (i4 == i4 || t4 == t4);
var l2 = { attribute: true, type: String, converter: o3, reflect: false, hasChanged: n3 };
var a2 = class extends HTMLElement {
  constructor() {
    super(), this._$Et = /* @__PURE__ */ new Map(), this.isUpdatePending = false, this.hasUpdated = false, this._$Ei = null, this.o();
  }
  static addInitializer(t4) {
    var i4;
    (i4 = this.l) !== null && i4 !== void 0 || (this.l = []), this.l.push(t4);
  }
  static get observedAttributes() {
    this.finalize();
    const t4 = [];
    return this.elementProperties.forEach((i4, s7) => {
      const e6 = this._$Eh(s7, i4);
      e6 !== void 0 && (this._$Eu.set(e6, s7), t4.push(e6));
    }), t4;
  }
  static createProperty(t4, i4 = l2) {
    if (i4.state && (i4.attribute = false), this.finalize(), this.elementProperties.set(t4, i4), !i4.noAccessor && !this.prototype.hasOwnProperty(t4)) {
      const s7 = typeof t4 == "symbol" ? Symbol() : "__" + t4, e6 = this.getPropertyDescriptor(t4, s7, i4);
      e6 !== void 0 && Object.defineProperty(this.prototype, t4, e6);
    }
  }
  static getPropertyDescriptor(t4, i4, s7) {
    return { get() {
      return this[i4];
    }, set(e6) {
      const r6 = this[t4];
      this[i4] = e6, this.requestUpdate(t4, r6, s7);
    }, configurable: true, enumerable: true };
  }
  static getPropertyOptions(t4) {
    return this.elementProperties.get(t4) || l2;
  }
  static finalize() {
    if (this.hasOwnProperty("finalized"))
      return false;
    this.finalized = true;
    const t4 = Object.getPrototypeOf(this);
    if (t4.finalize(), this.elementProperties = new Map(t4.elementProperties), this._$Eu = /* @__PURE__ */ new Map(), this.hasOwnProperty("properties")) {
      const t5 = this.properties, i4 = [...Object.getOwnPropertyNames(t5), ...Object.getOwnPropertySymbols(t5)];
      for (const s7 of i4)
        this.createProperty(s7, t5[s7]);
    }
    return this.elementStyles = this.finalizeStyles(this.styles), true;
  }
  static finalizeStyles(i4) {
    const s7 = [];
    if (Array.isArray(i4)) {
      const e6 = new Set(i4.flat(1 / 0).reverse());
      for (const i5 of e6)
        s7.unshift(S2(i5));
    } else
      i4 !== void 0 && s7.push(S2(i4));
    return s7;
  }
  static _$Eh(t4, i4) {
    const s7 = i4.attribute;
    return s7 === false ? void 0 : typeof s7 == "string" ? s7 : typeof t4 == "string" ? t4.toLowerCase() : void 0;
  }
  o() {
    var t4;
    this._$Ep = new Promise((t5) => this.enableUpdating = t5), this._$AL = /* @__PURE__ */ new Map(), this._$Em(), this.requestUpdate(), (t4 = this.constructor.l) === null || t4 === void 0 || t4.forEach((t5) => t5(this));
  }
  addController(t4) {
    var i4, s7;
    ((i4 = this._$Eg) !== null && i4 !== void 0 ? i4 : this._$Eg = []).push(t4), this.renderRoot !== void 0 && this.isConnected && ((s7 = t4.hostConnected) === null || s7 === void 0 || s7.call(t4));
  }
  removeController(t4) {
    var i4;
    (i4 = this._$Eg) === null || i4 === void 0 || i4.splice(this._$Eg.indexOf(t4) >>> 0, 1);
  }
  _$Em() {
    this.constructor.elementProperties.forEach((t4, i4) => {
      this.hasOwnProperty(i4) && (this._$Et.set(i4, this[i4]), delete this[i4]);
    });
  }
  createRenderRoot() {
    var t4;
    const s7 = (t4 = this.shadowRoot) !== null && t4 !== void 0 ? t4 : this.attachShadow(this.constructor.shadowRootOptions);
    return i2(s7, this.constructor.elementStyles), s7;
  }
  connectedCallback() {
    var t4;
    this.renderRoot === void 0 && (this.renderRoot = this.createRenderRoot()), this.enableUpdating(true), (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
      var i4;
      return (i4 = t5.hostConnected) === null || i4 === void 0 ? void 0 : i4.call(t5);
    });
  }
  enableUpdating(t4) {
  }
  disconnectedCallback() {
    var t4;
    (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
      var i4;
      return (i4 = t5.hostDisconnected) === null || i4 === void 0 ? void 0 : i4.call(t5);
    });
  }
  attributeChangedCallback(t4, i4, s7) {
    this._$AK(t4, s7);
  }
  _$ES(t4, i4, s7 = l2) {
    var e6, r6;
    const h4 = this.constructor._$Eh(t4, s7);
    if (h4 !== void 0 && s7.reflect === true) {
      const n7 = ((r6 = (e6 = s7.converter) === null || e6 === void 0 ? void 0 : e6.toAttribute) !== null && r6 !== void 0 ? r6 : o3.toAttribute)(i4, s7.type);
      this._$Ei = t4, n7 == null ? this.removeAttribute(h4) : this.setAttribute(h4, n7), this._$Ei = null;
    }
  }
  _$AK(t4, i4) {
    var s7, e6, r6;
    const h4 = this.constructor, n7 = h4._$Eu.get(t4);
    if (n7 !== void 0 && this._$Ei !== n7) {
      const t5 = h4.getPropertyOptions(n7), l5 = t5.converter, a4 = (r6 = (e6 = (s7 = l5) === null || s7 === void 0 ? void 0 : s7.fromAttribute) !== null && e6 !== void 0 ? e6 : typeof l5 == "function" ? l5 : null) !== null && r6 !== void 0 ? r6 : o3.fromAttribute;
      this._$Ei = n7, this[n7] = a4(i4, t5.type), this._$Ei = null;
    }
  }
  requestUpdate(t4, i4, s7) {
    let e6 = true;
    t4 !== void 0 && (((s7 = s7 || this.constructor.getPropertyOptions(t4)).hasChanged || n3)(this[t4], i4) ? (this._$AL.has(t4) || this._$AL.set(t4, i4), s7.reflect === true && this._$Ei !== t4 && (this._$E_ === void 0 && (this._$E_ = /* @__PURE__ */ new Map()), this._$E_.set(t4, s7))) : e6 = false), !this.isUpdatePending && e6 && (this._$Ep = this._$EC());
  }
  async _$EC() {
    this.isUpdatePending = true;
    try {
      await this._$Ep;
    } catch (t5) {
      Promise.reject(t5);
    }
    const t4 = this.scheduleUpdate();
    return t4 != null && await t4, !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    var t4;
    if (!this.isUpdatePending)
      return;
    this.hasUpdated, this._$Et && (this._$Et.forEach((t5, i5) => this[i5] = t5), this._$Et = void 0);
    let i4 = false;
    const s7 = this._$AL;
    try {
      i4 = this.shouldUpdate(s7), i4 ? (this.willUpdate(s7), (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
        var i5;
        return (i5 = t5.hostUpdate) === null || i5 === void 0 ? void 0 : i5.call(t5);
      }), this.update(s7)) : this._$EU();
    } catch (t5) {
      throw i4 = false, this._$EU(), t5;
    }
    i4 && this._$AE(s7);
  }
  willUpdate(t4) {
  }
  _$AE(t4) {
    var i4;
    (i4 = this._$Eg) === null || i4 === void 0 || i4.forEach((t5) => {
      var i5;
      return (i5 = t5.hostUpdated) === null || i5 === void 0 ? void 0 : i5.call(t5);
    }), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t4)), this.updated(t4);
  }
  _$EU() {
    this._$AL = /* @__PURE__ */ new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$Ep;
  }
  shouldUpdate(t4) {
    return true;
  }
  update(t4) {
    this._$E_ !== void 0 && (this._$E_.forEach((t5, i4) => this._$ES(i4, this[i4], t5)), this._$E_ = void 0), this._$EU();
  }
  updated(t4) {
  }
  firstUpdated(t4) {
  }
};
a2.finalized = true, a2.elementProperties = /* @__PURE__ */ new Map(), a2.elementStyles = [], a2.shadowRootOptions = { mode: "open" }, h2 == null || h2({ ReactiveElement: a2 }), ((s3 = globalThis.reactiveElementVersions) !== null && s3 !== void 0 ? s3 : globalThis.reactiveElementVersions = []).push("1.1.0");

// node_modules/lit-element/lit-element.js
var l3;
var o4;
var s4 = class extends a2 {
  constructor() {
    super(...arguments), this.renderOptions = { host: this }, this._$Dt = void 0;
  }
  createRenderRoot() {
    var t4, e6;
    const i4 = super.createRenderRoot();
    return (t4 = (e6 = this.renderOptions).renderBefore) !== null && t4 !== void 0 || (e6.renderBefore = i4.firstChild), i4;
  }
  update(t4) {
    const i4 = this.render();
    this.hasUpdated || (this.renderOptions.isConnected = this.isConnected), super.update(t4), this._$Dt = x(i4, this.renderRoot, this.renderOptions);
  }
  connectedCallback() {
    var t4;
    super.connectedCallback(), (t4 = this._$Dt) === null || t4 === void 0 || t4.setConnected(true);
  }
  disconnectedCallback() {
    var t4;
    super.disconnectedCallback(), (t4 = this._$Dt) === null || t4 === void 0 || t4.setConnected(false);
  }
  render() {
    return b;
  }
};
s4.finalized = true, s4._$litElement$ = true, (l3 = globalThis.litElementHydrateSupport) === null || l3 === void 0 || l3.call(globalThis, { LitElement: s4 });
var n4 = globalThis.litElementPolyfillSupport;
n4 == null || n4({ LitElement: s4 });
((o4 = globalThis.litElementVersions) !== null && o4 !== void 0 ? o4 : globalThis.litElementVersions = []).push("3.1.0");

// node_modules/lit/node_modules/@lit/reactive-element/css-tag.js
var t3 = window.ShadowRoot && (window.ShadyCSS === void 0 || window.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype;
var e4 = Symbol();
var n5 = /* @__PURE__ */ new Map();
var s5 = class {
  constructor(t4, n7) {
    if (this._$cssResult$ = true, n7 !== e4)
      throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t4;
  }
  get styleSheet() {
    let e6 = n5.get(this.cssText);
    return t3 && e6 === void 0 && (n5.set(this.cssText, e6 = new CSSStyleSheet()), e6.replaceSync(this.cssText)), e6;
  }
  toString() {
    return this.cssText;
  }
};
var o5 = (t4) => new s5(typeof t4 == "string" ? t4 : t4 + "", e4);
var i3 = (e6, n7) => {
  t3 ? e6.adoptedStyleSheets = n7.map((t4) => t4 instanceof CSSStyleSheet ? t4 : t4.styleSheet) : n7.forEach((t4) => {
    const n8 = document.createElement("style"), s7 = window.litNonce;
    s7 !== void 0 && n8.setAttribute("nonce", s7), n8.textContent = t4.cssText, e6.appendChild(n8);
  });
};
var S3 = t3 ? (t4) => t4 : (t4) => t4 instanceof CSSStyleSheet ? ((t5) => {
  let e6 = "";
  for (const n7 of t5.cssRules)
    e6 += n7.cssText;
  return o5(e6);
})(t4) : t4;

// node_modules/lit/node_modules/@lit/reactive-element/reactive-element.js
var s6;
var e5 = window.trustedTypes;
var r5 = e5 ? e5.emptyScript : "";
var h3 = window.reactiveElementPolyfillSupport;
var o6 = { toAttribute(t4, i4) {
  switch (i4) {
    case Boolean:
      t4 = t4 ? r5 : null;
      break;
    case Object:
    case Array:
      t4 = t4 == null ? t4 : JSON.stringify(t4);
  }
  return t4;
}, fromAttribute(t4, i4) {
  let s7 = t4;
  switch (i4) {
    case Boolean:
      s7 = t4 !== null;
      break;
    case Number:
      s7 = t4 === null ? null : Number(t4);
      break;
    case Object:
    case Array:
      try {
        s7 = JSON.parse(t4);
      } catch (t5) {
        s7 = null;
      }
  }
  return s7;
} };
var n6 = (t4, i4) => i4 !== t4 && (i4 == i4 || t4 == t4);
var l4 = { attribute: true, type: String, converter: o6, reflect: false, hasChanged: n6 };
var a3 = class extends HTMLElement {
  constructor() {
    super(), this._$Et = /* @__PURE__ */ new Map(), this.isUpdatePending = false, this.hasUpdated = false, this._$Ei = null, this.o();
  }
  static addInitializer(t4) {
    var i4;
    (i4 = this.l) !== null && i4 !== void 0 || (this.l = []), this.l.push(t4);
  }
  static get observedAttributes() {
    this.finalize();
    const t4 = [];
    return this.elementProperties.forEach((i4, s7) => {
      const e6 = this._$Eh(s7, i4);
      e6 !== void 0 && (this._$Eu.set(e6, s7), t4.push(e6));
    }), t4;
  }
  static createProperty(t4, i4 = l4) {
    if (i4.state && (i4.attribute = false), this.finalize(), this.elementProperties.set(t4, i4), !i4.noAccessor && !this.prototype.hasOwnProperty(t4)) {
      const s7 = typeof t4 == "symbol" ? Symbol() : "__" + t4, e6 = this.getPropertyDescriptor(t4, s7, i4);
      e6 !== void 0 && Object.defineProperty(this.prototype, t4, e6);
    }
  }
  static getPropertyDescriptor(t4, i4, s7) {
    return { get() {
      return this[i4];
    }, set(e6) {
      const r6 = this[t4];
      this[i4] = e6, this.requestUpdate(t4, r6, s7);
    }, configurable: true, enumerable: true };
  }
  static getPropertyOptions(t4) {
    return this.elementProperties.get(t4) || l4;
  }
  static finalize() {
    if (this.hasOwnProperty("finalized"))
      return false;
    this.finalized = true;
    const t4 = Object.getPrototypeOf(this);
    if (t4.finalize(), this.elementProperties = new Map(t4.elementProperties), this._$Eu = /* @__PURE__ */ new Map(), this.hasOwnProperty("properties")) {
      const t5 = this.properties, i4 = [...Object.getOwnPropertyNames(t5), ...Object.getOwnPropertySymbols(t5)];
      for (const s7 of i4)
        this.createProperty(s7, t5[s7]);
    }
    return this.elementStyles = this.finalizeStyles(this.styles), true;
  }
  static finalizeStyles(i4) {
    const s7 = [];
    if (Array.isArray(i4)) {
      const e6 = new Set(i4.flat(1 / 0).reverse());
      for (const i5 of e6)
        s7.unshift(S3(i5));
    } else
      i4 !== void 0 && s7.push(S3(i4));
    return s7;
  }
  static _$Eh(t4, i4) {
    const s7 = i4.attribute;
    return s7 === false ? void 0 : typeof s7 == "string" ? s7 : typeof t4 == "string" ? t4.toLowerCase() : void 0;
  }
  o() {
    var t4;
    this._$Ep = new Promise((t5) => this.enableUpdating = t5), this._$AL = /* @__PURE__ */ new Map(), this._$Em(), this.requestUpdate(), (t4 = this.constructor.l) === null || t4 === void 0 || t4.forEach((t5) => t5(this));
  }
  addController(t4) {
    var i4, s7;
    ((i4 = this._$Eg) !== null && i4 !== void 0 ? i4 : this._$Eg = []).push(t4), this.renderRoot !== void 0 && this.isConnected && ((s7 = t4.hostConnected) === null || s7 === void 0 || s7.call(t4));
  }
  removeController(t4) {
    var i4;
    (i4 = this._$Eg) === null || i4 === void 0 || i4.splice(this._$Eg.indexOf(t4) >>> 0, 1);
  }
  _$Em() {
    this.constructor.elementProperties.forEach((t4, i4) => {
      this.hasOwnProperty(i4) && (this._$Et.set(i4, this[i4]), delete this[i4]);
    });
  }
  createRenderRoot() {
    var t4;
    const s7 = (t4 = this.shadowRoot) !== null && t4 !== void 0 ? t4 : this.attachShadow(this.constructor.shadowRootOptions);
    return i3(s7, this.constructor.elementStyles), s7;
  }
  connectedCallback() {
    var t4;
    this.renderRoot === void 0 && (this.renderRoot = this.createRenderRoot()), this.enableUpdating(true), (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
      var i4;
      return (i4 = t5.hostConnected) === null || i4 === void 0 ? void 0 : i4.call(t5);
    });
  }
  enableUpdating(t4) {
  }
  disconnectedCallback() {
    var t4;
    (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
      var i4;
      return (i4 = t5.hostDisconnected) === null || i4 === void 0 ? void 0 : i4.call(t5);
    });
  }
  attributeChangedCallback(t4, i4, s7) {
    this._$AK(t4, s7);
  }
  _$ES(t4, i4, s7 = l4) {
    var e6, r6;
    const h4 = this.constructor._$Eh(t4, s7);
    if (h4 !== void 0 && s7.reflect === true) {
      const n7 = ((r6 = (e6 = s7.converter) === null || e6 === void 0 ? void 0 : e6.toAttribute) !== null && r6 !== void 0 ? r6 : o6.toAttribute)(i4, s7.type);
      this._$Ei = t4, n7 == null ? this.removeAttribute(h4) : this.setAttribute(h4, n7), this._$Ei = null;
    }
  }
  _$AK(t4, i4) {
    var s7, e6, r6;
    const h4 = this.constructor, n7 = h4._$Eu.get(t4);
    if (n7 !== void 0 && this._$Ei !== n7) {
      const t5 = h4.getPropertyOptions(n7), l5 = t5.converter, a4 = (r6 = (e6 = (s7 = l5) === null || s7 === void 0 ? void 0 : s7.fromAttribute) !== null && e6 !== void 0 ? e6 : typeof l5 == "function" ? l5 : null) !== null && r6 !== void 0 ? r6 : o6.fromAttribute;
      this._$Ei = n7, this[n7] = a4(i4, t5.type), this._$Ei = null;
    }
  }
  requestUpdate(t4, i4, s7) {
    let e6 = true;
    t4 !== void 0 && (((s7 = s7 || this.constructor.getPropertyOptions(t4)).hasChanged || n6)(this[t4], i4) ? (this._$AL.has(t4) || this._$AL.set(t4, i4), s7.reflect === true && this._$Ei !== t4 && (this._$E_ === void 0 && (this._$E_ = /* @__PURE__ */ new Map()), this._$E_.set(t4, s7))) : e6 = false), !this.isUpdatePending && e6 && (this._$Ep = this._$EC());
  }
  async _$EC() {
    this.isUpdatePending = true;
    try {
      await this._$Ep;
    } catch (t5) {
      Promise.reject(t5);
    }
    const t4 = this.scheduleUpdate();
    return t4 != null && await t4, !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    var t4;
    if (!this.isUpdatePending)
      return;
    this.hasUpdated, this._$Et && (this._$Et.forEach((t5, i5) => this[i5] = t5), this._$Et = void 0);
    let i4 = false;
    const s7 = this._$AL;
    try {
      i4 = this.shouldUpdate(s7), i4 ? (this.willUpdate(s7), (t4 = this._$Eg) === null || t4 === void 0 || t4.forEach((t5) => {
        var i5;
        return (i5 = t5.hostUpdate) === null || i5 === void 0 ? void 0 : i5.call(t5);
      }), this.update(s7)) : this._$EU();
    } catch (t5) {
      throw i4 = false, this._$EU(), t5;
    }
    i4 && this._$AE(s7);
  }
  willUpdate(t4) {
  }
  _$AE(t4) {
    var i4;
    (i4 = this._$Eg) === null || i4 === void 0 || i4.forEach((t5) => {
      var i5;
      return (i5 = t5.hostUpdated) === null || i5 === void 0 ? void 0 : i5.call(t5);
    }), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t4)), this.updated(t4);
  }
  _$EU() {
    this._$AL = /* @__PURE__ */ new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$Ep;
  }
  shouldUpdate(t4) {
    return true;
  }
  update(t4) {
    this._$E_ !== void 0 && (this._$E_.forEach((t5, i4) => this._$ES(i4, this[i4], t5)), this._$E_ = void 0), this._$EU();
  }
  updated(t4) {
  }
  firstUpdated(t4) {
  }
};
a3.finalized = true, a3.elementProperties = /* @__PURE__ */ new Map(), a3.elementStyles = [], a3.shadowRootOptions = { mode: "open" }, h3 == null || h3({ ReactiveElement: a3 }), ((s6 = globalThis.reactiveElementVersions) !== null && s6 !== void 0 ? s6 : globalThis.reactiveElementVersions = []).push("1.1.0");

export {
  $,
  y,
  b,
  w,
  o2 as o,
  r2 as r,
  s4 as s
};
/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
/**
 * @license
 * Copyright 2019 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
